Feature/6 category comment api

42jeelee opened the pull request:
### 카테고리 및 댓글 API 구현

#### 카테고리 API ( `/categories` )
> 카테고리 조회 시 해당 카테고리에 포함된 POST 개수를 같이 반환합니다.
```java
// code
Long postCount = countPostByCategoryId(categoryEntity.getCategoryId());

// CategoryRepository
@Query(value = "SELECT COUNT(p) FROM POST p WHERE p.category.categoryId = :id")
Long countPostByCategoryId(Long id);
```
- GET `/` : 전체 카테고리 조회
- GET `/{categoryId}` : 카테고리 조회


#### 댓글 API ( `/posts/{postId}/comments` )
- GET `/` : 전체 댓글 조회
- POST `/` : 댓글 달기
- DELETE `/{commentId}` : 댓글 삭제

Commit:
Merge branch 'feature/4-create-entity' into feature/6-category-comment-api

Commit:
ADD: Add Category API
- Add extends BaseTimeEntity (Category, Comment)
- Add Controller, Service for Category
- Add CategoryId in PostEntity

- Need HttpMessageNotReadableException
- record keyword

42jeelee committed:
ADD: Add Comment API
- Add Controller, Service for Comment

- Need fix foreign key all entitys
- Need handle HttpMessageNotReadableException
- record keyword

Commit:
FEAT: ADD category and comment api
- Add OneToMany annotation post, category entity
- Remove cascade ManyToOne annotation
- Fix errorCode at CommentService.deleteCommentById

- Need handle HttpMessageNotReadableException
- record keyword

42jeelee committed:
ADD: create data.sql for initial database
- Add config JPA in application.yaml
- Add data.sql

- Need handle HttpMessageNotReadableException
- record keyword

42jeelee committed:
FEAT: Update category url endpoint
- Change endpoint from /category to /categories

- Need handle HttpMessageNotReadableException
- record keyword

rarelhw reviewed (commented):
이전 풀리퀘가 머지 되면 diff 보도록 하겠습니다. 혹은 base 브랜치를 이전 풀리퀘의 브랜치로 잡으셔도 좋을 것 같습니다.

42jeelee committed:
FIX: Fix JPA cascade options
- Add Comments at PostEntity
- Add Posts at CategoryEntity
- remove async at PostService

- Need handle HttpMessageNotReadableException
- record keyword

42jeelee commented:
이전 #5 과 수정사항 커밋을 `git cherry-pick` 을 통해 병합했습니다.

rarelhw commented:
변경점이 21개나 되자나~ 징징댈꺼야.. ㅜㅜ 퇴근도 못했는데..
https://github.com/42jeelee/demoBoard-server/pull/5 여기서 본 코드 중복되니까 안볼래염.. 

![단비꺼야](https://github.com/user-attachments/assets/6338b133-a188-4596-af3a-996b419ba909)

이전 풀리퀘 머지해부러!

42jeelee committed:
Merge branch 'main' into feature/6-category-comment-api

42jeelee committed:
Feat: Override CORS configuration method in WebConfig

- Need handle HttpMessageNotReadableException
- record keyword

rarelhw commented on src/main/java/kr/co/jeelee/demoboard/domain/category/service/CategoryServiceImpl.java:
```diff
@@ -0,0 +1,41 @@
+		return categoryRepository.findAll().stream()
+			.map(categoryEntity -> {
+				Long postCount = countPostByCategoryId(categoryEntity.getCategoryId());
+				return CategoryResponse.of(categoryEntity, postCount);
```
카테고리를 가져올 때 해당 카테고리 내에 게시글의 수를 같이 보여주는 스팩이군요. 
이런 api는 요청은 요청 올때마다 매번 count 를 해야 하기 때문에 성능에 영향을 미칠 수 있어요.
스팩이 이렇다면 OK 지만, 실무에서 이렇게 만들면 전체적인 api의 성능을 저해할 수 있는 코드입니다. 

이런 경우 카테고리가 진짜 필요한 경우와 필요하지 않은 경우 api 에서 파라메터를 includesCounts 등의 값을 받아서 얘가 true인 경우에만 count 를 내려주게 하는 것도 방법이에요.

rarelhw commented on src/main/java/kr/co/jeelee/demoboard/domain/category/service/CategoryServiceImpl.java:
```diff
@@ -0,0 +1,41 @@
+	public CategoryResponse findById(Long id) {
+		return categoryRepository.findById(id)
+			.map(categoryEntity -> {
+				Long postCount = countPostByCategoryId(categoryEntity.getCategoryId());
```
루프 안에서 count 쿼리가 카테고리 갯수만큼 날라가네요.. 이러면 성능이.. 진짜 매우 안좋아져요.. 
DB 에 count 하는 쿼리를 루프 돌면서 날린다? 얘가 대용량 데이터 들어간 DB면 DBA 형이 몽둥이 들고 쫓아와요.

어차피 전체를 다 가져올라면 차라리..

select categoryId, count(1) from post group by categoryId 해온 결과를
associateBy(categoryId)하면 Map<categoryId, count> 이런 형태로 map 으로 만들어서.. 
코틀린은 list 에다가 associateBy 하면 Map으로 쇽샥 하고 바꿔주는데 최근 자바는 모르겄다.. 아마 최신 자바는 스트림에 이렇게 쇽샥 예쁘게 컬렉션 변환해주는거 있을껄?

쿼리는 한번만 날리고 루프 밖에서 map에 있는 값을 짝짝꿍 해서 데이터를 만드는 쪽이 성능이 더 좋아요.

rarelhw commented on src/main/java/kr/co/jeelee/demoboard/domain/category/service/CategoryServiceImpl.java:
```diff
@@ -0,0 +1,41 @@
+	}
+
+	@Override
+	public Long countPostByCategoryId(Long id) {
```
이 method 는 왜 public 으로 뽑았어요? 다른 패키지에서 얘 호출해요? 내부에서만 쓸꺼면 private 으로만 선언해도 될꺼 같은데용?

rarelhw commented on src/main/java/kr/co/jeelee/demoboard/domain/comment/service/CommentServiceImpl.java:
```diff
@@ -0,0 +1,57 @@
+        PostEntity postEntity = postRepository.findById(postId)
+                .orElseThrow(() -> new CustomException(ErrorCode.POST_NOT_FOUND));
+
+        CommentEntity commentEntity = CommentEntity.of(request.getAuthor(), request.getContent(), postEntity);
```
코드 깔끔하고 이쁘따 ^^

rarelhw reviewed (commented):
카테고리가 영 맘에 안드네용..

Commit:
Refactor: Add feature to toggle post count in categories based on query parameter
- Add `postCount` param at `/categories`
- Change all entity id name -> id

42jeelee commented:
### 리펙토링했습니다!
- **[Category post count 여부](https://github.com/42jeelee/demoBoard-server/pull/7#discussion_r1812452924)**
  + `/repositories` 의 파라미터로 `postCount` 값을 `true`로 설정하면 postCount를 추가로 받을 수 있습니다.
- **[roop안의 쿼리](https://github.com/42jeelee/demoBoard-server/pull/7#discussion_r1812459623)**
  + 하나의 쿼리로 categoryId 를 기준으로 left join을 통해 해당 post의 개수를 받습니다.
- **[내부 메서드 접근자](https://github.com/42jeelee/demoBoard-server/pull/7#discussion_r1812462105)**
  + 해당 메서드가 필요없어져 제거하였습니다.
- 이름 설정
  + 각 엔티티의 외래키 이름을 명시해주었습니다.
  + 각 엔티티의 기존 ID명 변경 `엔티티명Id` -> `id`

위에서 요구하신 것을 바탕으로 리펙토링하였습니다!

42jeelee committed:
Refactor: Change DTO to record
- Add PostService interface

- Need handle HttpMessageNotReadableException

rarelhw commented:
> ### 리펙토링했습니다!
> * **[Category post count 여부](https://github.com/42jeelee/demoBoard-server/pull/7#discussion_r1812452924)**
>   
>   * `/repositories` 의 파라미터로 `postCount` 값을 `true`로 설정하면 postCount를 추가로 받을 수 있습니다.
> * **[roop안의 쿼리](https://github.com/42jeelee/demoBoard-server/pull/7#discussion_r1812459623)**
>   
>   * 하나의 쿼리로 categoryId 를 기준으로 left join을 통해 해당 post의 개수를 받습니다.
> * **[내부 메서드 접근자](https://github.com/42jeelee/demoBoard-server/pull/7#discussion_r1812462105)**
>   
>   * 해당 메서드가 필요없어져 제거하였습니다.
> * 이름 설정
>   
>   * 각 엔티티의 외래키 이름을 명시해주었습니다.
>   * 각 엔티티의 기존 ID명 변경 `엔티티명Id` -> `id`
> 
> 위에서 요구하신 것을 바탕으로 리펙토링하였습니다!

roop -> Loop .... 'ㅅ'

rarelhw commented on src/main/java/kr/co/jeelee/demoboard/domain/category/service/CategoryServiceImpl.java:
```diff
@@ -0,0 +1,69 @@
+
+		return categoryRepository.findAll().stream()
+			.map(categoryEntity -> {
+				Long postCount = isPostCount
+					? countPostByCategoryId.get(categoryEntity.getId())
+					: null;
+
+				return CategoryResponse.of(categoryEntity, postCount);
```
위에서 countPostByCategoryId 가 isPostCount 에 의해서 이미 판단이 한번 됐기 때문에 뒤에서 또 삼항 연잔사 쓸 필요 없을 것 같아요.
어차피 false 면 null 일꺼라서.. 한줄로 가능할꺼 같아요!
```suggestion
				return CategoryResponse.of(categoryEntity, countPostByCategoryId.get(categoryEntity.getId()));
```

rarelhw commented on src/main/java/kr/co/jeelee/demoboard/domain/category/service/CategoryServiceImpl.java:
```diff
@@ -0,0 +1,69 @@
+					? countPostByCategoryId.get(categoryEntity.getId())
+					: null;
+
+				return CategoryResponse.of(categoryEntity, postCount);
```
요기도 마찬가지

rarelhw reviewed (approved):
좋은거 같아요!

Commit:
Refactor: Add countPostByCategoryId(id)
- Refactor CategoryServiceImpl Methods

 - Need handle HttpMessageNotReadableException

42jeelee committed:
Refactor: Add GlobalResponseAdvice for exception handling
- Add PostService interface

- Need handle HttpMessageNotReadableException
